Use dune to build unikernels

TheLortex opened the pull request:
This is hopefully the final attempt for the integration of `dune` in Mirage. There are various documents scattered around this repository explaining the changes. For reference:
- PR#1: https://github.com/mirage/mirage/pull/979
- PR#2:https://github.com/mirage/mirage/pull/1020
- PR#3: https://github.com/mirage/mirage/pull/1024
- PR#4: https://github.com/mirage/mirage/pull/1153
- Tracking issue: https://github.com/mirage/mirage/issues/1217

You should be able to test this PR using the following documentation: https://next.mirage.io/wiki/mirage-4
It uses https://github.com/mirage/mirage-dev/tree/mirage-4 for packages that haven't been upstreamed yet.


## Goal

The goal is to locally fetch unikernel dependencies and use `dune` to build unikernels, taking advantage of the `dune-workspace` (https://dune.readthedocs.io/en/stable/dune-files.html#dune-workspace) file to define and customize mirage targets. 

Only build tools are to be installed in the opam switch, the unikernel runtime dependencies are locally fetched using the `opam-monorepo` tool: https://github.com/ocamllabs/opam-monorepo. `opam-monorepo` lacks some features needed for mirage but they will be upstreamed and a release will be made before Mirage 4 comes out: in the meantime the required patches have been aggregated in my personal repository: https://github.com/TheLortex/duniverse/tree/mirage-4. 

## PR changes

I've split the changes into small commits for increased readability, but the commits are not self-contained (i.e. build might fail between two commits).

* **Functoria: add Action.run_cmd_cli**: When `config.exe` or `dune` is run it's nice to keep the interactive terminal capabilities. So the `Action.run_cmd_cli` action behaves similarly to `Action.run_cmd` but without touching `stdin` or `stdout`. 

* **Functoria: separate installation kind for packages**: Build tools need to be installed in the switch (ocaml-freestanding, solo5, ..) and runtime libraries to be locally fetched (mirage-solo5, mirage-runtime, ...), so we separate the two kind of packages using a `Global | Local` variant. Two opam files are generated, one consumed by `opam` and the other consumed by `opam-monorepo`. 

* **Functoria CLI: add extra-repo option**: Not all packages build with `dune`. For the others, an overlay opam repository has to be used in order to make them compatible with mirage. The `--extra-repo` option takes care of adding a custom overlay repository when fetching the dependencies. By default it's `https://github.com/mirage/opam-overlays`

* **Functoria: file generation only happen configure-time**:  At build time, a functoria device can use `dune` rules to generate files.

* **Functoria: move context cache in `<build_dir>/ <P.name> /context`**: Instead of `mirage.context`: the goal is to reduce the amount of files generated in the unikernel folder. For that matter, most files will be generated in  `<build_dir>/mirage`

* **Functoria: makefile -> use opam-monorepo to locally fetch deps**: Update Makefile generation

* **Functoria: add Dune helper module to generate dune files / devices can have dune rules to generate files at build time / generate dune files, use dune to build unikernels**: Generating dune files to build applications.

* **Functoria: update app_info device**: adapt to previous changes

* **Mirage: adapt to functoria changes**

* **Mirage: target interface changes**: the interface changes a bit, now that targets have to issue dune rules and workspace configuration to build unikernels

* **Mirage: adapt targets to dune**: target-specific rules to build unikernels using dune

* **Mirage: port devices to dune**: devices such as `crunch` now issues dune rules instead of build-time actions



Now that the dune build layout is approved and tested, it believe it's important to have it merged so that people other than me can get used to it and be able to contribute directly. If you have any remarks on any part of the changes I'll glad to hear them and make appropriate updates.

samoht commented on lib/functoria/cli.ml:
```diff
@@ -34,20 +34,24 @@ let description_section = "DESCRIBE OPTIONS"
-    ("install", `Install);
-    ("files-configure", `Files `Configure);
-    ("files-build", `Files `Build);
+    ("local.opam", `Opam `Local);
```
instead of `local` / `global` maybe name these with the intended goal? `build` / `release`  ?

TheLortex replied:
The thing is you can have "build" tools such as `crunch` or ppx's that are locally fetched in your unikernel workspace, so that might be confusing. The technical requirements are:
- `local`: builds with dune (either runtime libraries or build-time tools)
- `global`: only build-time tools

Drup commented:
Is that the final final one? :)

I'm not completely sure about having dune stanza in functoria devices, but maybe tying functoria with dune is the best way forward. It certainly seems like a working solution! For the rest, it looks good to me.

dinosaure commented:
Can you rebase your PR with `master`? It includes the patch to fix OCaml-CI (with `alcotest.1.4.0`).

Drup commented:
I'm not opposed to it, on the contrary. My very initial worry was in the context where there was no "dominant" tool, and the most used one wasn't terribly satisfactory. This is not really the case anymore. :)

yomimono commented:
If there are specific tasks or tests required to help with merging and releasing this, please let me know. I find myself with (1) a bit of extra time and (2) unikernels I'd like to build within a larger Dune project, and (2) doesn't work well with the current released tooling.

samoht commented on CHANGES.md:
```diff
@@ -1,3 +1,35 @@
+
+Main changes:
+
+* `dune build` is used to build unikernels. Because dune builds
```
"because dune buils" ? seems that the end of the sentence is missing?

samoht commented on CHANGES.md:
```diff
@@ -1,3 +1,35 @@
+Main changes:
+
+* `dune build` is used to build unikernels. Because dune builds
+* Two opam files are generated configure-time:
```
```suggestion
* Two opam files are generated at configure-time:
```

samoht commented on CHANGES.md:
```diff
@@ -1,3 +1,35 @@
+* Two opam files are generated configure-time:
+  - `<unikernel>-install.opam`: for global dependencies that are to be installed with opam.
+  - `<unikernel>.opam`: for unikernel dependencies as they are to be locally fetched.
+* Unikernel dependencies are fetched in the source folder using the
+`opam-monorepo` tool, formerly known as `duniverse`. This tool reads the
+generated opam file and uses the opam solver to compute the transitive
+dependency set, then fetch sources in a `duniverse/` subfolder.
+More info on the [Github repository](https://github.com/ocamllabs/duniverse).
+Disclaimer: the project is unreleased and is subject to changes.
+* Solo5 provides a cross-compiler that is used to pass on the freestanding
+flags when building unikernel sources.
```
```suggestion
* Unikernel dependencies are fetched in the source folder using the
  `opam-monorepo` tool, formerly known as `duniverse`. This tool reads the
  generated opam file and uses the opam solver to compute the transitive
  dependency set, then fetch sources in a `duniverse/` subfolder.
  More info on the [Github repository](https://github.com/ocamllabs/opam-monorepo).
  Disclaimer: the project is unreleased and is subject to changes.

* Solo5 provides a cross-compiler that is used to pass on the freestanding
  flags when building unikernel sources.
```

samoht commented on lib/functoria/filegen.ml:
```diff
@@ -42,12 +43,15 @@ module Make (P : PROJECT) = struct
     | `Sexp -> Fmt.str ";; Generated by"
     | `Opam | `Make -> "# Generated by"
     | `OCaml -> "(* Generated by"
+    | `OpamLock -> "synopsis: \"duniverse generated"
```
why does `headers` and `short_headers` disagree on what to do for `OpamLock`?

samoht replied:
Also: `OpamLock` doesn't seem to be used elsewhere?

TheLortex replied:
`headers` is used when Mirage generates files.
`short_headers` is used to identify files generated by Mirage.

For the opam lockfile, only the identification function (`short_headers`) was used to remove the file in the `clean` step. Now that most mirage files (include the lockfile) are stored together inside the `mirage/` folder, there's no need to identify `OpamLock` anymore so it can be removed

samoht commented on lib/functoria/package.mli:
```diff
@@ -18,10 +18,17 @@
 (** Representation of opam packages. *)
 
+type kind = [ `Global | `Local ]
+(** The kind of package installation:
```
```suggestion
(** The scope of package installation:
```

samoht commented on lib/functoria/package.mli:
```diff
@@ -18,10 +18,17 @@
 (** The type for opam packages. *)
 
 val v :
+  ?kind:kind ->
```
I'd probably name this "scope" instead of "kind" (as I still find the local/global distinction a bit confusing), but that's super minor.

samoht commented on test/mirage/query/run.t:
```diff
@@ -4,73 +4,128 @@ Query name
   ]
   
 
+Query local opam
```
So by reading that example, I got the local/global distinction wrong. @TheLortex can you try to clarify this a bit?

What's the difference between the local/global/install/locked opam files? :-)

TheLortex replied:
There are three opam files in a mirage project:
- `<name>.opam`: unikernel depencies that have to be locally fetched so that they exist in the dune workspace. Notably all runtime dependencies should be in that opam file, but it can also contain build tools as long as they build using dune. 
- `<name>.opam.locked`: unikernel dependencies lockfile obtained using `opam-monorepo`. 
- `<name>-install.opam`: the rest of the build tools, the main example being `ocaml-freestanding`. They are installed in the switch using opam. 

I agree that the names are a bit confusing but I don't know any other option. The rationale is that there are dependencies that need to be installed _globally_ in your opam switch and that there are runtime dependencies that need to be _locally_ fetched alongside the unikernel files as a monorepo in a dune workspace.

TheLortex replied:
New proposal: 
* _global_ becomes _switch_
* _local_ becomes _monorepo_
Therefore the names are more in sync with the traditional ocaml concepts (opam switch and opam-monorepo)

What do you think ?

dinosaure commented:
> If there are specific tasks or tests required to help with merging and releasing this, please let me know. I find myself with (1) a bit of extra time and (2) unikernels I'd like to build within a larger Dune project, and (2) doesn't work well with the current released tooling.

Currently, we have an unofficial explanation from @TheLortex about MirageOS 4 which is available here: https://next.mirage.io/wiki/mirage-4. It will be nice if you can reproduce the process locally and give to us some feedbacks 👍 - it will help us a lot to catch missing steps/states/etc.

yomimono commented:
With a couple [small](https://github.com/mirage/mirage-www/pull/702) [changes](https://github.com/mirage/cowabloga/pulls) I was able to build mirage-www on OCaml 4.12 with the mirage-4 development repository. I'm able to run the unikernel via hvt as in the simple instructions. :tada:

TheLortex commented:
Some updates on the PR:
- Improve the changelog description
- I've renamed the opam file distinction `local`/`global` to `monorepo`/`switch` to improve clarity
- `Package.kind` is now `Package.scope`
- Now that opam 2.1 is out, I've removed the 2.0 support to keep the generated `Makefile` simple. 
- Removed the `run ['_build/default/./config.exe' ...]: exited with 1` message when a command would fail

Mirage CI is still using 2.0 so I'll update that

samoht commented:
Can you rebase/cleanup against master? I think it's good to merge.

I have extra changes to run e2e tests as cram tests, I can add those after the merge to avoid more conflicts.

samoht commented:
> even if you lost the history

Yea it was done on purpose, as the various commits didn't compile separately and were intertwined :-)

It's also fine to make other PRs to fix the remaining issues (if needed)

samoht commented:
@TheLortex is the workflow to a PR documented somewhere? Do you think we should keep that workflow once 4.00 is released or it's just a temporary workflow? Maybe we want this to happen for every commit to `main`?

TheLortex commented:
The pipeline is described here: https://github.com/ocurrent/mirage-ci
I'm working on fixing it. I'll keep a MirageOS 3 pipeline to watch the `3` branch

TheLortex commented:
So to be clear, the CI tests this PR along with https://github.com/mirage/mirage-dev/pull/345 and https://github.com/mirage/mirage-skeleton/pull/303

dinosaure commented:
@samoht you can merge when you want, the CI is green 🌲

samoht commented:
Welcome to MirageOS 4.0 :p

smorimoto commented:
Woohoo!

yomimono commented:
HELL YEAH
